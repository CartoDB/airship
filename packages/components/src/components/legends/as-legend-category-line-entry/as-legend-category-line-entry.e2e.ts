import { E2EPage, newE2EPage } from '@stencil/core/testing';

describe('as-legend-category-line-entry', async () => {
  describe('Rendering', async () => {
    let page: E2EPage;

    it('should render label and color, with 1.5px solid style by default', async () => {
      page = await newE2EPage({ html: `
        <as-legend-category-line-entry>
        </as-legend-category-line-entry>
      ` });

      const element = await page.find('as-legend-category-line-entry');
      element.setProperty('label', 'Hello there');
      element.setProperty('color', '#FABADA');
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });

    it('should support custom stroke width', async () => {
      page = await newE2EPage({ html: `
        <as-legend-category-line-entry>
        </as-legend-category-line-entry>
      ` });

      const element = await page.find('as-legend-category-line-entry');
      element.setProperty('label', 'Hello there');
      element.setProperty('color', '#FABADA');
      element.setProperty('width', '10');
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });

    it('should support custom stroke style', async () => {
      page = await newE2EPage({ html: `
        <as-legend-category-line-entry>
        </as-legend-category-line-entry>
      ` });

      const element = await page.find('as-legend-category-line-entry');
      element.setProperty('label', 'Hello there');
      element.setProperty('color', '#FABADA');
      element.setProperty('strokeStyle', 'dotted');
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });
  });
});
