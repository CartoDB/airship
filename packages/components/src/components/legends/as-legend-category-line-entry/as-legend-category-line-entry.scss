@import '../../../../../styles/src/core/_dev';

as-legend-category-line-entry {
  .as-legend-category-line-entry--wrapper {
    display: flex;
    align-items: center;
  }

  .as-legend-category-line-entry--line {
    width: get-spacing(4);
    height: 0;
    margin: auto;
  }
}
