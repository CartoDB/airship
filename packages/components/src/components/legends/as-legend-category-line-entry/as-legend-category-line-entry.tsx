import { Component, Prop, State, Watch } from '@stencil/core';

const MAX_WIDTH = 16;

@Component({
  shadow: false,
  styleUrl: './as-legend-category-line-entry.scss',
  tag: 'as-legend-category-line-entry',
})
export class LegendCategoryLineEntry {
  @Prop() public label: string;
  @Prop() public color: string;
  @Prop() public strokeStyle: string = 'solid';
  @Prop() public width: number = 1.5;

  @State() private _width: number;

  public componentWillLoad() {
    this._widthChanged(this.width);
  }

  public render() {
    return (
      <div class='as-legend-category-line-entry--wrapper'>
        <div class='as-legend-category--figure--wrapper'>
          <div class='as-legend-category-line-entry--line' style={this.getStyle()}></div>
        </div>
        <span class='as-legend-category--label'>{this.label}</span>
      </div>
    );
  }

  @Watch('width')
  public _widthChanged(newValue: number) {
    this._width = Math.min(MAX_WIDTH, newValue);
  }

  private getStyle() {
    return {
      borderTopColor: this.color,
      borderTopStyle: this.strokeStyle,
      borderTopWidth: `${this._width}`
    };
  }
}
