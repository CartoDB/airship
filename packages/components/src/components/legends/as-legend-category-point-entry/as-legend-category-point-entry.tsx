import { Component, Prop } from '@stencil/core';
import { borderStyleCounts } from '../../../utils/border-style-counts';

// This component ignores the strokeWidth property, and always paints
// a 1px border.
const FAKE_BORDER_SIZE = 1;

const DEFAULT_WIDTH = 16;
// Required for background masks to work properly
const SYMBOL_OFFSET = 2;

@Component({
  shadow: false,
  styleUrl: './as-legend-category-point-entry.scss',
  tag: 'as-legend-category-point-entry',
})
export class LegendCategoryPointEntry {
  @Prop() public label: string;
  @Prop() public color: string;
  @Prop() public strokeColor: string;
  @Prop() public strokeStyle: string;
  @Prop() public marker: string;
  @Prop() public width: number = DEFAULT_WIDTH;

  public render() {
    const classes = {
      'as-legend-category-point-entry--point': !this.marker,
      'as-legend-category-point-entry--symbol': true
    };

    return (
      <div class='as-legend-category-point-entry--wrapper'>
        <div class='as-legend-category--figure--wrapper'>
          <div class={classes} style={this.getStyle()}></div>
        </div>
        <span class='as-legend-category--label'>{this.label}</span>
      </div>
    );
  }

  private getStyle() {
    return {
      backgroundColor: this.color,
      border: `${FAKE_BORDER_SIZE}px ${this.strokeStyle || 'solid'} ${this.strokeColor}`,
      ...this.getWidth(),
      ...this.getMask()
    };
  }

  private getWidth() {
    if (!this.width) {
      return {};
    }

    let sizeOffset = borderStyleCounts(this.strokeStyle)
      ? FAKE_BORDER_SIZE * 2
      : 0;

    if (this.marker) {
      sizeOffset += SYMBOL_OFFSET;
    }

    return {
      height: `${this.width + sizeOffset}px`,
      width: `${this.width + sizeOffset}px`
    };
  }

  private getMask() {
    if (!this.marker) {
      return {};
    }

    return {
      '-webkit-mask-image': `url(${this.marker})`,
      '-webkit-mask-position': 'center',
      '-webkit-mask-repeat': 'no-repeat',
      '-webkit-mask-size': `${this.width}px`,
      'maskImage': `url(${this.marker})`,
      'maskPosition': 'center',
      'maskRepeat': 'no-repeat',
      'maskSize': `${this.width}px`
    };
  }
}
