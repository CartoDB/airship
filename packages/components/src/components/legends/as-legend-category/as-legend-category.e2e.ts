import { E2EPage, newE2EPage } from '@stencil/core/testing';

describe('as-legend-category', async () => {
  describe('Rendering', async () => {
    let page: E2EPage;

    it('should render mixed types of legend entries', async () => {
      page = await newE2EPage({ html: `
        <as-legend-category>
        </as-legend-category>
      ` });

      const element = await page.find('as-legend-category');
      element.setProperty('data', [{
        color: '#FF0000',
        label: 'point label',
        strokeColor: '#00FF00',
        type: 'point',
        width: 16
      },
      {
        color: '#FF0000',
        label: 'line label',
        type: 'line'
      },
      {
        color: '#FF0000',
        label: 'polygon label',
        strokeColor: '#00FF00',
        type: 'line'
      }]);
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });

    describe('should define a CSS variable', () => {
      it('with the max width of its children', async () => {
        page = await newE2EPage({ html: `
          <as-legend-category>
          </as-legend-category>
        ` });

        const element = await page.find('as-legend-category');
        element.setProperty('data', [{
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 16
        },
        {
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 32
        },
        {
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 20
        }]);

        await page.waitForChanges();
        // CSS variables can't be accessed any other way right now
        expect(element.innerHTML).toContain('--as--basic--legend--figure-width: 32px');
      });

      it('but never below 16', async () => {
        page = await newE2EPage({ html: `
          <as-legend-category>
          </as-legend-category>
        ` });

        const element = await page.find('as-legend-category');
        element.setProperty('data', [{
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 10
        }]);

        await page.waitForChanges();
        // CSS variables can't be accessed any other way right now
        expect(element.innerHTML).toContain('--as--basic--legend--figure-width: 16px');
      });
    });
  });
});
