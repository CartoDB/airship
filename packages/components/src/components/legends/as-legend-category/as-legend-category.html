
<!DOCTYPE html>
<html lang="en">

<head>
    <base href="/">
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">

    <title>Category legend</title>

    <link rel="stylesheet" href="packages/styles/dist/airship.css">
    <link rel="stylesheet" href="packages/icons/dist/icons.css">
    <script type="module" src="packages/components/dist/airship/airship.esm.js"></script>
    <script nomodule="" src="packages/components/dist/airship/airship.js"></script>
    <script src="/packages/bridge/dist/asbridge.js"></script>

    <style>
      #legends {
        flex-direction: column;
        width: 100%;
      }
    </style>
</head>

<body class="as-app-body">
  <div class="as-app">
    <main class="as-main">
        <div class="as-map-area">
            <div id="map"></div>
            <div class="as-map-panels">
                <div class="as-panel as-panel--top as-panel--left" id="legends">

                  <!-- Point -->
                  <as-legend
                    id="legendsLocation"
                    heading="Point category">
                    <div slot="legends">
                      <as-legend-category
                        orientation="horizontal"
                        id="legendPointHorizontal">
                      </as-legend-category>
                      <as-legend-category
                        id="legendPoint">
                      </as-legend-category>
                    </div>
                  </as-legend>

                  <!-- Line -->
                  <as-legend
                    id="legendsLocation"
                    heading="Line category">
                    <div slot="legends">
                      <as-legend-category
                        orientation="horizontal"
                        id="legendLineHorizontal">
                      </as-legend-category>
                      <as-legend-category
                        id="legendLine">
                      </as-legend-category>
                    </div>
                  </as-legend>

                  <!-- Polygon -->
                  <as-legend
                    id="legendsLocation"
                    heading="Polygon category">
                    <div slot="legends">
                      <as-legend-category
                        orientation="horizontal"
                        id="legendPolygonHorizontal">
                      </as-legend-category>
                      <as-legend-category
                        id="legendPolygon">
                      </as-legend-category>
                    </div>
                  </as-legend>

                  <!-- Fun mix -->
                  <as-legend
                    id="legendsLocation"
                    heading="Mixed content">
                    <div slot="legends">
                      <as-legend-category
                        orientation="horizontal"
                        id="legendMixedHorizontal">
                      </as-legend-category>
                      <as-legend-category
                        id="legendMixed">
                      </as-legend-category>
                    </div>
                  </as-legend>
                </div>
            </div>
        </div>
    </main>
  </div>
  <script>
    // Point
    const point = legendPointHorizontal.data = legendPoint.data = [{
      type: 'point',
      width: 10,
      color: '#FF00FF',
      strokeColor: '#00FFFF',
      label: 'Label'
    },{
      type: 'point',
      width: 16,
      color: '#FF0000',
      strokeColor: '#00FF00',
      label: 'Longer label'
    },{
      type: 'point',
      width: 22,
      color: '#0000FF',
      strokeColor: '#00FFFF',
      label: 'Slightly longer label'
    },{
      type: 'point',
      width: 28,
      color: '#00FFFF',
      strokeColor: '#FFFFFF',
      label: 'Longish label'
    }];

    // Line
    const line = legendLineHorizontal.data = legendLine.data = [{
      type: 'line',
      width: 10,
      color: '#FF00FF',
      strokeColor: '#00FFFF',
      label: 'Label'
    },{
      type: 'line',
      width: 16,
      color: '#FF0000',
      strokeColor: '#00FF00',
      label: 'Longer label'
    },{
      type: 'line',
      width: 22,
      color: '#0000FF',
      strokeColor: '#00FFFF',
      label: 'Slightly longer label'
    },{
      type: 'line',
      width: 28,
      color: '#00FFFF',
      strokeColor: '#FFFFFF',
      label: 'Longish label'
    }];

    // Polygon
    const polygon = legendPolygonHorizontal.data = legendPolygon.data = [{
      type: 'polygon',
      color: '#FF00FF',
      label: '#FF00FF'
    },{
      type: 'polygon',
      color: '#FF0000',
      label: '#FF0000'
    },{
      type: 'polygon',
      color: '#0000FF',
      label: '#0000FF'
    },{
      type: 'polygon',
      color: '#00FFFF',
      label: '#00FFFF'
    }];

    legendMixedHorizontal.data = legendMixed.data = [].concat(point, line, polygon);
  </script>
</body>

</html>
