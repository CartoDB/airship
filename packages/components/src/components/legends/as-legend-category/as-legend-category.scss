@import '../../../../../styles/src/core/_dev';

as-legend-category {
  --as--basic--legend--font: var(--as--font--caption);
  --as--basic--legend--text-transform: capitalize;

  display: block;

  .as-legend-category--wrapper {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
  }

  .as-legend-category--label {
    font: var(--as--basic--legend--font);
    text-transform: var(--as--basic--legend--text-transform);
  }

  & .as-legend-category--figure--wrapper {
    width: var(--as--basic--legend--figure-width);
    margin-right: get-spacing(2);
  }

  .as-legend-category--wrapper-horizontal > .as-legend-category--entry {
    margin-right: get-spacing(3);
  }

  .as-legend-category--wrapper:not(.as-legend-category--wrapper-horizontal) > .as-legend-category--entry:not(:first-child) {
    margin-top: get-spacing(3);
  }

  .as-legend-category--wrapper-horizontal {
    flex-direction: row;
    align-items: center;
  }
}
