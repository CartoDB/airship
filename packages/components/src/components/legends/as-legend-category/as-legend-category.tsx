import { Component, h, Prop } from '@stencil/core';

const MIN_WIDTH = 16;

@Component({
  shadow: false,
  styleUrl: './as-legend-category.scss',
  tag: 'as-legend-category',
})
export class LegendColorCategory {
  @Prop() public data: LegendData[];
  @Prop() public orientation: 'horizontal' | 'vertical';
  @Prop() public width: number;

  public render() {
    if (!this.data || this.data.length === 0) {
      return null;
    }

    const wrapper = {
      'as-legend-category--wrapper': true,
      'as-legend-category--wrapper-horizontal': this.orientation === 'horizontal'
    };

    return <div style={this.getStyle()} class={wrapper}>
      {
        this.data
          .map((e) => this.renderLegend(e))
          .filter((e) => e !== null)
          .map((e) => <div class='as-legend-category--entry'>{e}</div>)
      }
    </div>;
  }

  private renderLegend(legend: LegendData) {
    switch (legend.type) {
      case 'point':
        return <as-legend-category-point-entry
          label={legend.label}
          width={this.width || legend.width}
          color={legend.color}
          strokeColor={legend.strokeColor}
          marker={legend.marker}
          strokeStyle={legend.strokeStyle}
          >
        </as-legend-category-point-entry>;
      case 'line':
        return <as-legend-category-line-entry
          label={legend.label}
          width={legend.width}
          color={legend.color}
          strokeStyle={legend.strokeStyle}
          >
        </as-legend-category-line-entry>;
      case 'polygon':
        return <as-legend-category-polygon-entry
          label={legend.label}
          color={legend.color}
          strokeColor={legend.strokeColor}
          strokeStyle={legend.strokeStyle}
          >
        </as-legend-category-polygon-entry>;
      default:
        return null;
    }
  }

  private getStyle() {
    const maxLegendWidth = this.data.slice().sort(
      (first, second) => second.width - first.width
    )[0].width;

    return {
      '--as--basic--legend--figure-width': `${this.width || Math.max(maxLegendWidth, MIN_WIDTH)}px`
    };
  }
}
