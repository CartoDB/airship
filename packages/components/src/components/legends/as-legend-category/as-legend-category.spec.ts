import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { LegendCategoryLineEntry } from '../as-legend-category-line-entry/as-legend-category-line-entry';
import { LegendCategoryPointEntry } from '../as-legend-category-point-entry/as-legend-category-point-entry';
import { LegendColorCategory } from './as-legend-category';

describe('as-legend-category', () => {
  describe('Rendering', () => {
    let page: SpecPage;

    it('should render mixed types of legend entries', async () => {
      page = await newSpecPage({
        components: [LegendColorCategory,
          LegendCategoryLineEntry,
          LegendCategoryPointEntry],
        html: `<as-legend-category></as-legend-category>`
      });

      const element = page.root;
      element.data = [{
        color: '#FF0000',
        label: 'point label',
        strokeColor: '#00FF00',
        type: 'point',
        width: 16
      },
      {
        color: '#FF0000',
        label: 'line label',
        type: 'line'
      },
      {
        color: '#FF0000',
        label: 'polygon label',
        strokeColor: '#00FF00',
        type: 'line'
      }];
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });

    describe('should define a CSS variable', () => {
      it('with the max width of its children', async () => {
        page = await newSpecPage({
          components: [LegendColorCategory],
          html: `<as-legend-category></as-legend-category>`
        });

        const element = page.root;
        element.data = [{
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 16
        },
        {
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 32
        },
        {
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 20
        }];

        await page.waitForChanges();
        // CSS variables can't be accessed any other way right now
        expect(element.innerHTML).toContain('--as--basic--legend--figure-width: 32px');
      });

      it('but never below 16', async () => {
        page = await newSpecPage({
          components: [LegendColorCategory],
          html: `<as-legend-category></as-legend-category>`
        });

        const element = page.root;
        element.data = [{
          color: '#FF0000',
          label: 'point label',
          strokeColor: '#00FF00',
          type: 'point',
          width: 10
        }];

        await page.waitForChanges();
        // CSS variables can't be accessed any other way right now
        expect(element.innerHTML).toContain('--as--basic--legend--figure-width: 16px');
      });
    });
  });
});
