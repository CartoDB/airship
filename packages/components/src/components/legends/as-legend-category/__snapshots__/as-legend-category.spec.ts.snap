// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-legend-category Rendering should render mixed types of legend entries 1`] = `"<as-legend-category><div class=\\"as-legend-category--wrapper\\" style=\\"--as--basic--legend--figure-width: 16px;\\"><div class=\\"as-legend-category--entry\\"><as-legend-category-point-entry><div class=\\"as-legend-category-point-entry--wrapper\\"><div class=\\"as-legend-category--figure--wrapper\\"><div class=\\"as-legend-category-point-entry--point as-legend-category-point-entry--symbol\\" style=\\"background-color: #FF0000; border: 1px solid #00FF00; height: 18px; width: 18px;\\"></div></div><span class=\\"as-legend-category--label\\">point label</span></div></as-legend-category-point-entry></div><div class=\\"as-legend-category--entry\\"><as-legend-category-line-entry><div class=\\"as-legend-category-line-entry--wrapper\\"><div class=\\"as-legend-category--figure--wrapper\\"><div class=\\"as-legend-category-line-entry--line\\" style=\\"border-top-color: #FF0000; border-top-style: solid; border-top-width: 1.5;\\"></div></div><span class=\\"as-legend-category--label\\">line label</span></div></as-legend-category-line-entry></div><div class=\\"as-legend-category--entry\\"><as-legend-category-line-entry><div class=\\"as-legend-category-line-entry--wrapper\\"><div class=\\"as-legend-category--figure--wrapper\\"><div class=\\"as-legend-category-line-entry--line\\" style=\\"border-top-color: #FF0000; border-top-style: solid; border-top-width: 1.5;\\"></div></div><span class=\\"as-legend-category--label\\">polygon label</span></div></as-legend-category-line-entry></div></div></as-legend-category>"`;
